### PR TITLE
add script to upgrade etcd aiding monitoring etcd

### PR DESCRIPTION
**Pull request by xudifsd**
Use readiness probe to monitor etcd service.

As long as pod with labels: app=name, monitored=pai(previous watchdog only requires app=name labels), the watchdog will monitor its readiness status, and generate corresponding metrics, and will sent out email if service is not ready for a long time.

This PR added all service with these two labels and add readiness probe to etcd service. For etcd service deployed by previous pai code, the PR has a script for user to upgrade. If users want to monitor old version etcd, they can refer the doc and do some operation.

I assume this PR can go into v0.11.0, and can modify upgrade doc according if this PR can not make it.

### CODE REVIEW

**Comment by hao1939** [comment]
Hi @xudifsd ,

Why we need two labels? `labels: app=name, monitored=pai`.

**Comment by fanyangCS** [comment]
what if the k8s is not deployed by ourselved. does the design still work?

File: docs/monitor-etcd.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Background
+
+In Pai version before 0.10.0, pai did not have a monitor solution for etcd deployed
+by OpenPai, so admin will not get notified when something wrong with etcd and can not
+running correctly.
+
+To monitor etcd, we have watchdog service, this service monitor pods with special
+labels and check its running status via readiness probe. If you have deployed
```
Comment (fanyangCS) [inline docs/monitor-etcd.md]: what is the value of the special label?

File: docs/monitor-etcd.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+before 0.10.0, you should do following upgrade to configure etcd correctly so it can
+be monitored.
+
+# Upgrade steps
+
+Please note following steps involves restart etcd pods, the etcd process will stop for around 5 minutes, and **kubernetes can not work during the time**, please plan upgrade in advance to minimize the possible impact.
+
+1. Please use a [dev-box](pai-management/doc/how-to-setup-dev-box.md) to do following operations, as dev-box can provide a clean environment.
+2. Checkout the 0.10.0 version branch of pai: `cd $PAI_ROOT && git checkout v0.10.0`.
+3. Apply upgrade script: `cd $PAI_ROOT/src/etcd-upgrade/build && kubectl apply -f ds.yaml`.
```
Comment (fanyangCS) [inline docs/monitor-etcd.md]: >ds.yaml [](start = 82, length = 7)

use a daemonSet to upgrade etcd is way too heavy. by the way, can we just label the pod? k8s does support attaching a label to a node on the fly, right?

File: docs/monitor-etcd.md (hunk 1 above)
Comment (xudifsd) [reply docs/monitor-etcd.md]: app=name, monitored=pai

File: docs/monitor-etcd.md (hunk 2 above)
Comment (xudifsd) [reply docs/monitor-etcd.md]: Yes, we are updating the etcd.yaml, by using DaemonSet to deploy update script in every node and update etcd.yaml.

**Comment by xudifsd** [comment]
> Hi @xudifsd ,
> 
> Why we need two labels? `labels: app=name, monitored=pai`.

Previous, I just use app=name, but this is too simple, and if user deploy other pods using app labels, their pods will be monitored by pai, which not what we expected. To use monitored=pai, this marks pod we want to monitored and is obvious for users.

**Comment by xudifsd** [comment]
related to #2126

**Comment by hao1939** [comment]
> > Hi @xudifsd ,
> > Why we need two labels? `labels: app=name, monitored=pai`.
> 
> Previous, I just use app=name, but this is too simple, and if user deploy other pods using app labels, their pods will be monitored by pai, which not what we expected. To use monitored=pai, this marks pod we want to monitored and is obvious for users.

Hi Di,

We should alter if user deploy other pods on the k8s, shouldn't we?
Running un-managed pod on the k8s will affect our stability.
At default, we should monitor everything, so we'd better add an exclusive rules instead of inclusive rules.

**Comment by xudifsd** [comment]
@hao1939 I am fine with it, just leave it as app=name. I thought since editing etc label is not easy, we should do it right, and do it explicitly with monitored=pai.

What's your opinion? @fanyangCS.

**Comment by hao1939** [comment]
Hi @xudifsd ,

As you said, edits labels on pod is not easy, could we put the rules inside the alter manager?
So we can avoid to touch the pod be monitored.

**Comment by xudifsd** [comment]
> Hi @xudifsd ,
> 
> As you said, edits labels on pod is not easy, could we put the rules inside the alter manager?
> So we can avoid to touch the pod be monitored.

I'm not sure what you mean, the pod needs to attached special labels to be found by watchdog, and the discovery rule is defined in watchdog, not in pod. And this PR is propose using rule `app=name && monitored=pai` to select pods.

**Comment by hao1939** [comment]
> > Hi @xudifsd ,
> > As you said, edits labels on pod is not easy, could we put the rules inside the alter manager?
> > So we can avoid to touch the pod be monitored.
> 
> I'm not sure what you mean, the pod needs to attached special labels to be found by watchdog, and the discovery rule is defined in watchdog, not in pod. And this PR is propose using rule `app=name && monitored=pai` to select pods.
I mean that we only need the minimum tags, for example `app=name`. Once we got the app name, we can easily decision whether to include/exclude it from monitoring.

**Comment by xudifsd** [comment]
@hao1939  removed monitored=pai rule

**Comment by fanyangCS** [comment]
@hao1939, make sure you can handle this new alert in AKS case.

**Comment by xudifsd** [comment]
AKS will not need to monitor etcd

**Pull request merged**